core/test/solver/batch_cg.cpp

```diff
@@ -0,0 +1,277 @@
+    ASSERT_EQ(copy->get_num_batch_items(), this->num_batch_items);
+    auto copy_mtx = gko::as<Solver>(copy.get())->get_system_matrix();
+    const auto copy_batch_mtx = gko::as<const Mtx>(copy_mtx.get());
+    GKO_ASSERT_BATCH_MTX_NEAR(this->mtx.get(), copy_batch_mtx, 0.0);
+}
+
+
+TYPED_TEST(BatchCg, CanBeMoved)
+{
+    using Mtx = typename TestFixture::Mtx;
+    using Solver = typename TestFixture::Solver;
+    auto copy = this->solver_factory->generate(Mtx::create(this->exec));
+
+    copy->move_from(this->solver);
+
+    ASSERT_EQ(copy->get_common_size(),
+              gko::dim<2>(this->num_rows, this->num_rows));
+    ASSERT_EQ(copy->get_num_batch_items(), this->num_batch_items);
+    auto copy_mtx = gko::as<Solver>(copy.get())->get_system_matrix();
+    const auto copy_batch_mtx = gko::as<const Mtx>(copy_mtx.get());
+    GKO_ASSERT_BATCH_MTX_NEAR(this->mtx.get(), copy_batch_mtx, 0.0);
+}
+
+
+TYPED_TEST(BatchCg, CanBeCloned)
+{
+    using Mtx = typename TestFixture::Mtx;
+    using Solver = typename TestFixture::Solver;
+
+    auto clone = this->solver->clone();
+
+    ASSERT_EQ(clone->get_common_size(),
+              gko::dim<2>(this->num_rows, this->num_rows));
+    ASSERT_EQ(clone->get_num_batch_items(), this->num_batch_items);
+    auto clone_mtx = gko::as<Solver>(clone.get())->get_system_matrix();
+    const auto clone_batch_mtx = gko::as<const Mtx>(clone_mtx.get());
+    GKO_ASSERT_BATCH_MTX_NEAR(this->mtx.get(), clone_batch_mtx, 0.0);
+}
+
+
+TYPED_TEST(BatchCg, CanBeCleared)
+{
+    using Solver = typename TestFixture::Solver;
+
+    this->solver->clear();
+
+    ASSERT_EQ(this->solver->get_num_batch_items(), 0);
+    auto solver_mtx = gko::as<Solver>(this->solver.get())->get_system_matrix();
+    ASSERT_EQ(solver_mtx, nullptr);
+}
+
+
+TYPED_TEST(BatchCg, CanSetCriteriaInFactory)
+{
+    using Solver = typename TestFixture::Solver;
+    using real_type = typename TestFixture::real_type;
+
+    auto solver_factory =
+        Solver::build()
+            .with_max_iterations(22)
+            .with_tolerance(static_cast<real_type>(0.25))
+            .with_tolerance_type(gko::batch::stop::tolerance_type::relative)
+            .on(this->exec);
+
+    auto solver = solver_factory->generate(this->mtx);
+    ASSERT_EQ(solver->get_parameters().max_iterations, 22);
+    ASSERT_EQ(solver->get_parameters().tolerance, 0.25);
+    ASSERT_EQ(solver->get_parameters().tolerance_type,
+              gko::batch::stop::tolerance_type::relative);
+}
+
+
+TYPED_TEST(BatchCg, CanSetResidualTol)
+{
+    using Solver = typename TestFixture::Solver;
+    using real_type = typename TestFixture::real_type;
+    auto solver_factory =
+        Solver::build()
+            .with_max_iterations(22)
+            .with_tolerance(static_cast<real_type>(0.25))
+            .with_tolerance_type(gko::batch::stop::tolerance_type::relative)
+            .on(this->exec);
+    auto solver = solver_factory->generate(this->mtx);
+
+    solver->reset_tolerance(0.5);
```

Comment:
If we already allow this (which I don't support), then it also needs to set the tolerance of the parameter. Having two different tolerances is nonsense to me.
(Same for the other reset functions)

Reply:
I think we decided to make the `parameters_` struct not public facing anymore, so in that case it should be irrelevant. Currently, we cannot set the values of the `parameters_` struct from the `reset_` functions because the `reset_` functions are part of the base class, where there is no concept of the parameters_ struct. In any case, the parameters_ struct is not used in the solver core anymore, with the updated SolverBase structure and only used to pass the parameters to the base solver members.

Reply:
sounds reasonable.